From 5e70c4a1c4b4dbc8cbb25a85595e0a9f42f3b12c Mon Sep 17 00:00:00 2001
From: mpolacek <mpolacek@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Thu, 16 Jul 2015 13:25:25 +0000
Subject: [PATCH] 	PR c++/66748 	* tree.c (handle_abi_tag_attribute):
 Check for CLASS_TYPE_P before 	accessing TYPE_LANG_SPECIFIC node.

	* g++.dg/abi/abi-tag15.C: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-5-branch@225877 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/cp/ChangeLog                     | 9 +++++++++
 gcc/cp/tree.c                        | 6 ++++--
 gcc/testsuite/ChangeLog              | 8 ++++++++
 gcc/testsuite/g++.dg/abi/abi-tag15.C | 3 +++
 4 files changed, 24 insertions(+), 2 deletions(-)
 create mode 100644 gcc/testsuite/g++.dg/abi/abi-tag15.C

diff --git a/gcc/cp/ChangeLog b/gcc/cp/ChangeLog
index 8522ae5..92d56e5 100644
--- a/gcc/cp/ChangeLog
+++ b/gcc/cp/ChangeLog
@@ -1,3 +1,12 @@
+2015-07-16  Marek Polacek  <polacek@redhat.com>
+
+	Backported from mainline
+	2015-07-08  Marek Polacek  <polacek@redhat.com>
+
+	PR c++/66748
+	* tree.c (handle_abi_tag_attribute): Check for CLASS_TYPE_P before
+	accessing TYPE_LANG_SPECIFIC node.
+
 2015-07-16  Release Manager
 
 	* GCC 5.2.0 released.
diff --git a/gcc/cp/tree.c b/gcc/cp/tree.c
index 3037564..f0a9589 100644
--- a/gcc/cp/tree.c
+++ b/gcc/cp/tree.c
@@ -3602,13 +3602,15 @@ handle_abi_tag_attribute (tree* node, tree name, tree args,
 		 name, *node);
 	  goto fail;
 	}
-      else if (CLASSTYPE_TEMPLATE_INSTANTIATION (*node))
+      else if (CLASS_TYPE_P (*node)
+	       && CLASSTYPE_TEMPLATE_INSTANTIATION (*node))
 	{
 	  warning (OPT_Wattributes, "ignoring %qE attribute applied to "
 		   "template instantiation %qT", name, *node);
 	  goto fail;
 	}
-      else if (CLASSTYPE_TEMPLATE_SPECIALIZATION (*node))
+      else if (CLASS_TYPE_P (*node)
+	       && CLASSTYPE_TEMPLATE_SPECIALIZATION (*node))
 	{
 	  warning (OPT_Wattributes, "ignoring %qE attribute applied to "
 		   "template specialization %qT", name, *node);
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 1d2cca3..45561a0 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,11 @@
+2015-07-16  Marek Polacek  <polacek@redhat.com>
+
+	Backported from mainline
+	2015-07-08  Marek Polacek  <polacek@redhat.com>
+
+	PR c++/66748
+	* g++.dg/abi/abi-tag15.C: New test.
+
 2015-07-16  Release Manager
 
 	* GCC 5.2.0 released.
diff --git a/gcc/testsuite/g++.dg/abi/abi-tag15.C b/gcc/testsuite/g++.dg/abi/abi-tag15.C
new file mode 100644
index 0000000..bfda3a2
--- /dev/null
+++ b/gcc/testsuite/g++.dg/abi/abi-tag15.C
@@ -0,0 +1,3 @@
+// PR c++/66748
+
+enum __attribute__((abi_tag("foo"))) E {}; // { dg-error "redeclaration of" }
-- 
1.9.4

